File: code_samples/section6/problems/problem6_2/problem6_2.js
/*
    This file demonstrates building a MIN-HEAP in-place from an array, then
    validating correctness by repeatedly extracting the minimum from a COPY.

    Min-heap invariant (0-based indexing):
        For every i > 0:
            arr[parent(i)] <= arr[i]

    Index formulas:
        parent(i) = Math.floor((i - 1) / 2)
        left(i)   = 2*i + 1
        right(i)  = 2*i + 2

    Main pieces:
      - heapify(arr, n, i): "sift down" (recursive) to fix subtree rooted at i
      - buildHeap(arr): bottom-up heap construction (O(n))
      - popMin(arr, sizeObj): test helper that pops min from heap prefix [0..size-1]
      - tests: print original array, build heap, extract sorted order

    Notes:
      - heapify is recursive. For very large heaps, an iterative version avoids deep recursion.
      - popMin returns null when empty (clearer than sentinel values).
*/

/*
    heapify (recursive sift-down)
    -----------------------------
    Ensures that the subtree rooted at index i satisfies the min-heap property,
    assuming the subtrees rooted at its children already satisfy the property.

    Parameters:
      arr : array storing heap elements
      n   : current heap size (only indices [0..n-1] are considered part of heap)
      i   : root index of the subtree to fix

    Algorithm:
      1) Find the smallest among i, left child, right child (if children exist).
      2) If a child is smaller than arr[i], swap with the smallest child.
      3) Recurse on the child position we swapped into (because that subtree may now be invalid).

    Complexity:
      O(log n) worst case (height of heap).
*/
function heapify(arr, n, i) {
    let smallest = i;
    const left = 2 * i + 1;
    const right = 2 * i + 2;

    // If left child exists and is smaller than current smallest, update candidate.
    if (left < n && arr[left] < arr[smallest]) smallest = left;

    // If right child exists and is smaller than current smallest, update candidate.
    if (right < n && arr[right] < arr[smallest]) smallest = right;

    // If the smallest isn't the current node, swap and continue heapifying downward.
    if (smallest !== i) {
        [arr[i], arr[smallest]] = [arr[smallest], arr[i]];
        heapify(arr, n, smallest);
    }
}

/*
    buildHeap
    ---------
    Converts an arbitrary array into a min-heap IN PLACE using bottom-up construction.

    Key insight:
      - Leaves (indices >= floor(n/2)) already satisfy the heap property (no children).
      - The last internal node is floor(n/2) - 1.
      - Heapify each internal node from that index down to 0.

    Complexity:
      O(n) overall.
*/
function buildHeap(arr) {
    const n = arr.length;

    // Start from the last internal node and sift down each node toward the root.
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }
}

// ===== Tests =====

/*
    printArray
    ----------
    Utility function to print a label + array contents.
    (When printing a heap, remember it's "heap order", not sorted order.)
*/
function printArray(label, arr) {
    console.log(label + arr.join(" "));
}

/*
    popMin
    ------
    Test helper: pops the minimum element from an array-based min-heap that
    lives in the prefix arr[0..size-1].

    Why sizeObj:
      - JavaScript numbers are passed by value, so we wrap "size" in an object
        to emulate pass-by-reference and allow popMin to decrement it.

    Parameters:
      arr     : array containing heap elements
      sizeObj : object like { size: <current heap size> }

    Returns:
      - null when heap is empty
      - otherwise the removed minimum value

    Steps:
      1) Save root (arr[0]).
      2) Decrease size.
      3) If still non-empty:
           - move last heap element into root
           - heapify down using the NEW size as n
*/
function popMin(arr, sizeObj) {
    let size = sizeObj.size;
    if (size === 0) return null;

    // Root holds the minimum in a min-heap.
    const root = arr[0];

    // Shrink the heap by one element.
    size--;
    sizeObj.size = size;

    if (size > 0) {
        // Move the last element in the heap prefix to the root position.
        // Then restore heap property within [0..size-1].
        arr[0] = arr[size];
        heapify(arr, size, 0);
    }

    return root;
}

// ===== Test run =====

/*
    1) Start with an unsorted array.
    2) buildHeap => transforms it to a min-heap in-place.
    3) Extract repeatedly from a copy => should yield ascending order.

    Expected sorted output for [9, 4, 7, 1, 0, 3, 2]:
        0 1 2 3 4 7 9
*/
let arr = [9, 4, 7, 1, 0, 3, 2];

printArray("Original array: ", arr);

buildHeap(arr);
printArray("After buildHeap (min-heap array): ", arr);

// Extract values to test correctness
console.log("Extracting elements in sorted order:");

// Clone heap since extraction mutates the heap structure.
let heapCopy = arr.slice();

// Track logical heap size separately as it shrinks on each pop.
let sizeObj = { size: heapCopy.length };

let out = [];
while (sizeObj.size > 0) {
    // popMin returns null only when empty; loop condition prevents null entries.
    out.push(popMin(heapCopy, sizeObj));
}

// out should be in ascending order if heap is valid.
printArray("", out);

File: code_samples/section6/problems/problem6_2/problem6_2.py
"""
Problem 6.2: Build a min-heap from an array (bottom-up heap construction).

This script includes:
  1) heapify(arr, n, i): recursive "sift down" to fix the subtree rooted at i
  2) build_heap(arr): in-place O(n) heap construction from an arbitrary list
  3) pop_min(arr, size_ref): test helper that pops min from heap prefix [0..size-1]
  4) A simple print-based test harness

Min-heap invariant (0-based indexing):
    For every i > 0:
        arr[parent(i)] <= arr[i]

Index formulas:
    parent(i) = (i - 1) // 2
    left(i)   = 2*i + 1
    right(i)  = 2*i + 2

Complexity:
  - build_heap: O(n) overall
  - heapify (sift-down): O(log n)
  - pop_min (test helper): O(log n)

Notes:
  - heapify is recursive; for very large heaps, an iterative version avoids deep recursion.
  - Your pop_min docstring says it returns int, but it can return None when empty.
    Using Optional[int] makes that explicit.
"""

from typing import Optional


def heapify(arr: list[int], n: int, i: int) -> None:
    """
    Heapify (sift down) the subtree rooted at index i to satisfy min-heap property.

    Parameters:
        arr: The list containing heap elements.
        n:   The current heap size (only arr[0..n-1] is considered part of heap).
        i:   The index of the subtree root to fix.

    Algorithm:
        1) Identify the smallest among i, left child, right child (if children exist).
        2) If a child is smaller than arr[i], swap with the smallest child.
        3) Recursively heapify the child index we swapped into.

    Complexity:
        O(log n) worst case (height of heap).
    """
    smallest = i
    left = 2 * i + 1
    right = 2 * i + 2

    # Compare against left child (if within heap prefix)
    if left < n and arr[left] < arr[smallest]:
        smallest = left

    # Compare against right child (if within heap prefix)
    if right < n and arr[right] < arr[smallest]:
        smallest = right

    # If the smallest is not the current node, swap and continue fixing downward
    if smallest != i:
        arr[i], arr[smallest] = arr[smallest], arr[i]
        heapify(arr, n, smallest)


def build_heap(arr: list[int]) -> None:
    """
    Build a min-heap in place from an arbitrary list.

    Bottom-up construction:
        - Leaves (indices >= n//2) are already heaps of size 1.
        - Start from the last internal node (n//2 - 1) and heapify each node
          moving backward to index 0.

    Parameters:
        arr: List to transform into a min-heap in place.

    Complexity:
        O(n) overall.
    """
    n = len(arr)
    start = n // 2 - 1  # last internal node

    for i in range(start, -1, -1):
        heapify(arr, n, i)


# ===== Tests =====

def print_array(label: str, arr: list[int]) -> None:
    """
    Print a label followed by the elements of arr separated by spaces.

    Note:
        When printing a heap array, the order shown is "heap order", not sorted order.
    """
    print(label + " ".join(str(x) for x in arr))


def pop_min(arr: list[int], size_ref: list[int]) -> Optional[int]:
    """
    Pop the min element from an array-based min-heap using prefix size.

    Why size_ref is a list:
        Python integers are immutable and passed by assignment. Using a single-element
        list emulates pass-by-reference so the function can decrement the heap size.

    Parameters:
        arr:      Heap array (elements are valid in prefix arr[0..size-1]).
        size_ref: Single-element list holding current heap size, i.e. [size].

    Returns:
        The minimum element (root) if heap is non-empty; otherwise None.

    Steps:
        1) If size == 0: return None.
        2) Save root = arr[0].
        3) Decrement size.
        4) If size > 0:
             - Move last element in prefix to root
             - Heapify down from root using the NEW size as n
        5) Return saved root.

    Complexity:
        O(log n) for non-empty heap.
    """
    size = size_ref[0]
    if size == 0:
        return None

    # Minimum is always at index 0 in a min-heap
    root = arr[0]

    # Shrink heap size
    size -= 1
    size_ref[0] = size

    if size > 0:
        # Move the last element of the active heap prefix to the root
        arr[0] = arr[size]

        # Restore heap property within prefix [0..size-1]
        heapify(arr, size, 0)

    return root


# ===== Run Tests =====

array = [9, 4, 7, 1, 0, 3, 2]

print_array("Original array: ", array)

# Transform list into min-heap in place
build_heap(array)

print_array("After build_heap (min-heap array): ", array)

# Validate correctness by extracting all elements (should be ascending)
print("Extracting elements in sorted order:")
heap_copy = array[:]  # clone heap because popping mutates the list contents/structure
size_r = [len(heap_copy)]  # mutable "reference" to logical heap size

output: list[int] = []
while size_r[0] > 0:
    v = pop_min(heap_copy, size_r)
    # v will not be None here because loop condition ensures size_r[0] > 0,
    # but we keep the Optional return type for correctness and safety.
    output.append(v)  # type: ignore[arg-type]

print_array("", output)
